File: Sum of Subsets.cpp
#include<iostream>
#include<algorithm>
using namespace std;

int n,m,w[100],res[100]={0};

void printArray(int arr[],int n)
{
   for(int i=0;i<n;i++)
    {
        cout<<arr[i]<<" ";
    }
    cout<<endl<<endl;
}

void SumofSubsets(int s, int k, int r)
{
    res[k]=1;

    if(s+w[k]==m)
    {
        //cout<<"Final Result = "<<endl;
        printArray(res,k+1);
    }
    else if(s+w[k]+w[k+1] <= m)
        SumofSubsets(s+w[k], k+1, r-w[k]);

    if(((s+r-w[k])>=m) && ((s+w[k+1])<=m) )
    {
        res[k]=0;
        SumofSubsets(s,k+1,r-w[k]);
    }
}

int main()
{
    int sum=0;

    cout<<"Enter the number of elements : "<<endl;
    cin>>n;


    cout<<"Enter the weight required weight of subsets : "<<endl;
    cin>>m;

    cout<<"Enter the elements one by one : "<<endl;
    for(int i=0;i<n;i++)
    {
        cout<<"w["<<i<<"] = ";
        cin>>w[i];
        sum=sum+w[i];
    }
    cout<<endl;

    cout<<"Entered elements are : "<<endl;
    printArray(w,n);

    sort(w,w+n);
    cout<<"Elements after sorting are : "<<endl;
    printArray(w,n);

    cout<<"Total Weight of the array : "<<sum<<endl;

    SumofSubsets(0,0,sum);


    return 0;
}

File: Sum of Subsets.py
#Try all the comments to understand the working of the program
#Bakctracking!


global n,m,x,w
def SumofSub(s,k,r):
    x[k]=1
    #print("First x in function = ",x)
    #print()
    if(s+w[k]==m):
        print(x[:k+1])
        #print("s = ",s," w[k] = ",w[k]," m = ",m)
        #print()
    elif ((s+w[k]+w[k+1])<=m):
        #print("In Elif condition")
        #print("s = ",s," w[k] = ",w[k],"w[k+1] = ",w[k+1], "m = ",m)
        #print("s+w[k]+w[k+1] = ",s+w[k]+w[k+1], "m = ",m)
        #print()
        SumofSub(s+w[k],k+1,r-w[k])

    if ( ((s+r-w[k])>=m) and ((s+w[k+1])<=m)):
        
        x[k]=0
        #print("x in Second if is ",x)
        #print("s = ",s," r = ",r," w[k] = ",w[k], "w[k+1] = ",w[k+1])
        #print("s+r-w[k] = ",s+r-w[k],"s+w[k+1] = ",s+w[k+1]," m = ",m)
        #print()
        SumofSub(s,k+1,r-w[k])

# n is number of elements
n=int(input("Enter the nunber of elements : "))
# m is required sum of subset
m=int(input("Enter the weight required by the subsets : "))
# w is weights of n elements
print("Enter the elements in a single line separated by a space : ")
w=list(map(int,input().split()))
w.sort()
# solution vector
x=[]
#Initialize solution vector x[0:n-1]=[0..0]
for i in range(n):
    x.append(0)
#print("x = ",x)
#print()
SumofSub(0,0,sum(w))
